File: 백준/Gold/2631. 줄세우기/줄세우기.py
import sys
input = sys.stdin.readline

n = int(input())

lis = []
for i in range(n):
    num = int(input())
    start = 0
    end = len(lis) - 1

    if end == -1:
        lis.append(num)
        continue

    if lis[-1] < num:
        lis.append(num)
        continue
    
    while start <= end:
        mid = (start + end) // 2
        if lis[mid] >= num:
            end = mid - 1
        else:
            start = mid + 1

    if start >= len(lis):
        lis.append(num)
    else:
        lis[start] = num

print(n - len(lis))

File: 백준/Gold/2631. 줄세우기/줄세우기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine()); // 입력 개수
        List<Integer> lis = new ArrayList<>(); // LIS를 저장할 리스트

        for (int i = 0; i < n; i++) {
            int num = Integer.parseInt(br.readLine()); // 입력 값
            int start = 0;
            int end = lis.size() - 1;

            // 리스트가 비어 있는 경우, 값을 추가
            if (lis.isEmpty() || lis.get(lis.size() - 1) < num) {
                lis.add(num);
                continue;
            }

            // 이진 탐색으로 적절한 위치를 찾기
            while (start <= end) {
                int mid = (start + end) / 2;
                if (lis.get(mid) >= num) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }

            // 찾은 위치에 현재 값을 삽입
            lis.set(start, num);
        }

        // 결과 출력
        System.out.println(n - lis.size());
    }
}
